watcher/local/statespubsub.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2021 - See NOTICE file for copyright holders.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package local
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"perun.network/go-perun/channel"
+	"perun.network/go-perun/pkg/sync/atomic"
+	"perun.network/go-perun/watcher"
+)
+
+var _ watcher.StatesPub = &statesPubSub{}
+
+// errAlreadyClosed indicates the pub-sub has already been closed.
+var errAlreadyClosed error = errors.New("already closed")
+
+const statesPubSubBufferSize = 10
+
+type (
+	statesPubSub struct {
+		closed atomic.Bool
+		pipe   chan channel.Transaction
+	}
+
+	// statesSub is used by the watcher to receive transactions.
+	statesSub interface {
+		statesStream() <-chan channel.Transaction
+		error() error
+		close() error
+	}
+)
+
+func newStatesPubSub() *statesPubSub {
+	return &statesPubSub{
+		pipe: make(chan channel.Transaction, statesPubSubBufferSize),
+	}
+}
+
+// Publish publishes the given state to all the subscribers.
+//
+// Each time when a transaction is published, watcher will treat it as the
+// latest transaction without any validation. It is the responsibility of the
+// client to publish transactions in correct order and ensure they are the
+// valid.
+//
+// Context is added as this method implements watcher.StatesPub interface.
+// However, is is not used as there no network communications involved in local
+// watcher.
+//
+// Returns nil if the state is published. Panics if the subscriptions is
+// already closed.
+func (s *statesPubSub) Publish(_ context.Context, tx channel.Transaction) error {
+	s.pipe <- tx
+	return nil
+}
+
+// close closes the publisher instance and all the subscriptions associated
+// with it. Any further call to Publish should panic.
```

Comment:
Suggestion: Use sync.Once.Do to close the subscription. The function should not panic. No error return value needed.

Question: Is there a benefit to throwing an error or panicing if you call the function twice?

Reply:
Yes, the benefit is much simpler implementation of pub-sub. Also, there is no reason for the Publish to be called after channel is closed. See [the above comment.](https://github.com/hyperledger-labs/go-perun/pull/212#discussion_r717660310)

Reply:
To your suggestion, **short answer** is yes, we can use `sync.Once` and remove error return. I have updated.

**Explanation**:
As, I mentioned in the conversation with @ggwpez earlier, I used `atomic.Bool` to track if a channel is closed and return error in two cases:
1. When `close` is called repeatedly.

    In this case, I realized it is acceptable if close is called multiple times. Only expectation from the caller is that the channel is closed after the `close` call returns. It is relevant if it was closed already or closed as a result of that call.

   Hence, removed error return for close.

2. When `Err` is called after the subscription is closed.

    In this case, the only way a local subscription is closed, is when the client calls `StopWatching` for the channel. After `StopWatching` is called, the `Watcher` will close the subscription. Earlier, we returned a `errAlreadyClosed` error in this case. But, I realized, it is more appropriate to return `nil`, because the subscription was closed only on the request of the client and there was no error encountered while closing it.

    Hence, updated `Err` to always return nil and documented the reason for the same.